### ISSUE TITLE
Decide on an approach for shared Aggregated API Server storage

### ISSUE THREAD

**Issue by fabiand** [kubernetes/kubernetes#46351]
User API Server (AS) becomes the more sophisticated alternative to TPRs.

While looking at the [example](https://github.com/kubernetes/sample-apiserver/blob/master/pkg/registry/wardle/etcd.go) we noticed that etcd was used as a backing store.

Now, it would be very convenient for plain TPRs could be used for storing the UAS data.
It would effectively allow us to stay within the boundaries of K8s, and not worry about coming up with our own etcd cluster or another storage method.

Thus my request boils down to having a simple way of using CDRs as storage for UAS.

**Comment by deads2k** [comment]
I don't think that CRDs are a proper fit for UAS storage.   CRDs are very opinionated about the shape of the data in order to conform with kube-like APIs.  Storage for a UAS must not have the very protections that CRD provides.  The first few that come to mind.

 1. CRD can only support JSON encoding since the data must be introspected.  UAS-storage must be unopinionated to support protobuf or other encodings.
 2. CRD can only support a single groupversion serialization to properly support kube versioning semantics: https://github.com/kubernetes/community/blob/master/contributors/design-proposals/thirdpartyresources.md#desired-api-semantics.  UAS-storage must support multiple groupversion serializations to support server-side schema evolution.
 3. CRD must enforce API semantics for update values.  UAS-storage must support storage semantics which are different.  For instance, DeletionTimestamp can only be set when an object is deleted in a REST API, but storage can convert deletes to updates.

I think that these differences are irreconcilable.  We want a different resource, with significantly different API semantics.  I'm in favor of creating such a resource, probably even in the same `apiextensions.k8s.io` API group, but I do not think that CRD is the answer.

@kubernetes/sig-api-machinery-misc

**Comment by deads2k** [comment]
@ncdc @smarterclayton to expand on the short conversation in slack.  The API semantics are very different.

**Comment by fabiand** [comment]
Right. Even if TPRs are not suited, it would still be good if the K8s API provides the required storage backend for UAS. As said initially, to me this would be very valuable, as we can rely on K8s for storage, and don't need to maintain our own datastore.

**Comment by fabiand** [comment]
> We want a different resource, with significantly different API semantics. I'm in favor of creating such a resource, probably even in the same

Could you explain this sentence actually a little?

**Comment by deads2k** [comment]
> Right. Even if TPRs are not suited, it would still be good if the K8s API provides the required storage backend for UAS. As said initially, to me this would be very valuable, as we can rely on K8s for storage, and don't need to maintain our own datastore.

I absolutely agree, I just want to be sure that everyone understands the shortcomings of CRDs for this application.

**Comment by deads2k** [comment]
> We want a different resource, with significantly different API semantics. I'm in favor of creating such a resource, probably even in the same
> Could you explain this sentence actually a little?

In order to create a storage API, the API doesn't end up looking like a kube-like API because of differences in introspection (cannot introspect storage), validation (cannot validate storage), update limitations (no update restrictions in storage), and delete semantics (delete always deletes).

In addition, you probably want the API to appear more like an etcd API.  I would actually consider something more along the lines of subresources on an instance of an `apistorage.apiextensions.k8s.io` resource.

**Comment by smarterclayton** [comment]
And before we clone the etcd API we need to have a larger discussion
about how this fits into the arc.  Aggregated API servers should be
easy, but don't have to be as easy as crd.

**Comment by fabiand** [comment]
Right - I feared we were not on the same side, but now it looks like we are.

**Comment by pmorie** [comment]
@deads2k what you say here makes sense to me: https://github.com/kubernetes/kubernetes/issues/46351#issuecomment-303760049

...but people are going to try to do this anyway, since they don't have a storage group yet.

That aside, do we have any plans to implement a UAS storage resource in 1.8?  I think it would be in our best interests to do so to avoid proliferation if people mis-using CRD to back a UAS.

**Comment by smarterclayton** [comment]
I would say my default position is that we will *not* create a generic etcd like storage api, because at best we would have a lowest common denominator etcd api that has none of the benefits and lots of disadvantages.

**Comment by deads2k** [comment]
> I would say my default position is that we will not create a generic etcd like storage api, because at best we would have a lowest common denominator etcd api that has none of the benefits and lots of disadvantages.

Do you have a suggestion for an alternative?  Providing a safe way to create a controlled proxy for the etcd API seems like the most reasonable option.  Just throwing out an option: create a resource in one location which surfaces as a functioning proxy on a subresource for said resource.

**Comment by smarterclayton** [comment]
etcd3 has greatly improved their grpcproxy functionality to handle many of
these cases - we could conceivably make that easier to use as part of these
cases.

**Comment by fabiand** [comment]
The issue with running any type of custom storage is, that the application needs to worry - and thus know! - about this storage method on Kubernetes.
It would be much more convenient if there was a way to leave this (object store) responsibility on Kubernetes.

**Comment by deads2k** [comment]
> etcd3 has greatly improved their grpcproxy functionality to handle many of
> these cases - we could conceivably make that easier to use as part of these
> cases.

You'd still need a resource to gate access to it and automatically provision it in a way that the user cannot gain general access.  I'm not very familiar with grpc, but why couldn't that exposed proxy be exposed via a subresource?

**Comment by bgrant0607** [comment]
Hmm. We need to decide what we're doing with respect to exposing the cluster state storage API.

https://github.com/kubernetes/kubernetes/issues/1957
https://github.com/kubernetes/features/issues/172

**Comment by fabiand** [comment]
Picking up @deads2k initial comment again. I understand that UAS can provide a wider range of API endpoints and formats. But on the other hand I would assume that it will be used to have something like what CRDs provide, just with validation and authorization. Thus I'd also expect that it will be "simple" objects (json/yaml based) which follow the general "rules" of CRDs. In those cases, I suppose, CRDs could be a good way to store the data of a UAS.

I.e in Kubevirt we could do this. (We actually do it today, by having our own custom API server for validation, and then store the data in CRDs).

**Comment by fabiand** [comment]
Is there any progress on or solution for this?

We are currently implementing a custom API server for KubeVirt kubevirt/kubevirt#355 and this issue becomes more pressing, as I don't want us to start relying on etcd for storing data.

**Comment by sttts** [comment]
> Is there any progress on or solution for this?

No progress unfortunately. As the time is limited until freeze (Sept 1), I don't see anything in 1.8. If we want that we have to start with a proposal now, to be implemented in 1.9.

**Comment by bgrant0607** [comment]
"User API server" is a misnomer. API aggregation is primarily intended to make it easier for us to build a non-monolithic system, to start new projects in separate repos (e.g., in kubernetes-incubator), to build extensions, and to reduce the need to fork the codebase for major extensions like Openshift.

We need a single storage API for all API servers, both the nucleus/core one and all others. We only want to solve performance, reliability, security, and semantic problems once. We need to be able to easily move API implementations between components as we carve the system into layers, integrate incubator projects, etc.

I don't think we know what that API looks like yet. Etcd3 added a lot of features for us (https://github.com/kubernetes/kubernetes/issues/1957#issuecomment-69863343) that we have yet to take advantage of, and can't until we complete the migration from etcd2. I have planned from the beginning of the project to break the resource representation into multiple key-value pairs and for the primary storage location to be indexed by uid rather than by name (name would become a symlink). Mitigating false conflicts (#8625) would be another thing to consider in the future.

For now, cluster administrators can either run a separate etcd3 instance for each set of apiserver replicas or configure them to talk to the same etcd3 instance, if they prefer.

**Comment by lavalamp** [comment]
This doc has mostly been approved by the SIG: https://docs.google.com/document/d/1i0xzRFB-uGLmLYueLMBTpHrOot9ScFxpkkcVcZHVbyA/edit#

(@smarterclayton and myself may hash out some of the language still, but it is close)

TL;DR: **Use etcd-operator to run your own etcd cluster.** If you can sweet talk the cluster admin into using the cluster's main etcd installation, great, but don't count on that. See the doc for a diagram.

**Comment by mswilson** [comment]
> This doc has mostly been approved by the SIG: https://docs.google.com/document/d/1i0xzRFB-uGLmLYueLMBTpHrOot9ScFxpkkcVcZHVbyA/edit#
> 
> (@smarterclayton and myself may hash out some of the language still, but it is close)

This doc is not public.

**Comment by lavalamp** [comment]
The doc is shared with the api machinery mailing list, kubernetes-announce, and kubernetes-dev, all of which anyone can join. (Sorry, my employer doesn't allow 100% open docs)

I thought we had exported it to a .md somewhere already though, but maybe not.
